ibm/service/power/resource_ibm_pi_volume_bulk.go

```diff
@@ -0,0 +1,524 @@
+}
+
+func resourceIBMPIVolumeBulkRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	sess, err := meta.(conns.ClientSession).IBMPISession()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	idArr, err := flex.IdParts(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	cloudInstanceID := idArr[0]
+	client := instance.NewIBMPIVolumeClient(ctx, sess, cloudInstanceID)
+	volumeIDs := idArr[1:]
+
+	d.Set(Arg_CloudInstanceID, cloudInstanceID)
+
+	// Set Arguments all volumes should have the same information
+	// so just get it from one
+	firstVolume, err := client.Get(volumeIDs[0])
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	d.Set(Arg_VolumePool, firstVolume.VolumePool)
+	if firstVolume.Shareable != nil {
+		d.Set(Arg_VolumeShareable, firstVolume.Shareable)
+	}
+	d.Set(Arg_VolumeSize, firstVolume.Size)
+	d.Set(Arg_VolumeType, firstVolume.DiskType)
+	d.Set(Arg_ReplicationEnabled, firstVolume.ReplicationEnabled)
+
+	result := make([]map[string]interface{}, 0, len(volumeIDs))
+	for _, volumeID := range volumeIDs {
+		vol, err := client.Get(volumeID)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		l := map[string]interface{}{
+			Attr_CRN:                  vol.Crn,
+			Attr_Auxiliary:            vol.Auxiliary,
+			Attr_AuxiliaryVolumeName:  vol.AuxVolumeName,
+			Attr_ConsistencyGroupName: vol.ConsistencyGroupName,
+			Attr_GroupID:              vol.GroupID,
+			Attr_IOThrottleRate:       vol.IoThrottleRate,
+			Attr_MasterVolumeName:     vol.MasterVolumeName,
+			Attr_MirroringState:       vol.MirroringState,
+			Attr_PrimaryRole:          vol.PrimaryRole,
+			Attr_ReplicationSites:     vol.ReplicationSites,
+			Attr_ReplicationStatus:    vol.ReplicationStatus,
+			Attr_ReplicationType:      vol.ReplicationType,
+			Attr_VolumeStatus:         vol.State,
+			Attr_WWN:                  vol.Wwn,
+		}
+		if vol.DeleteOnTermination != nil {
+			l[Attr_DeleteOnTermination] = vol.DeleteOnTermination
+		}
+		if vol.VolumeID != nil {
+			l[Attr_VolumeID] = vol.VolumeID
+		}
+		result = append(result, l)
+	}
+
+	d.Set(Attr_Volumes, result)
+
+	return nil
+}
+
+func resourceIBMPIVolumeBulkDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	sess, err := meta.(conns.ClientSession).IBMPISession()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	idArr, err := flex.IdParts(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	cloudInstanceID := idArr[0]
+	client := instance.NewIBMPIVolumeClient(ctx, sess, cloudInstanceID)
+
+	volumeIDs := idArr[1:]
+	volumesDelete := models.VolumesDelete{
+		VolumeIDs: volumeIDs,
+	}
+
+	volInfo, err := client.BulkVolumeDelete(&volumesDelete)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	log.Printf("[DEBUG] Volumes delete accepted: %s", volInfo.Summary)
+
+	_, err = isWaitForIBMPIVolumeBulkDeleted(ctx, client, volumeIDs, d.Timeout(schema.TimeoutDelete))
+	if err != nil {
+		d.SetId(cloudInstanceID + err.Error())
```

Comment:
Not sure why we are adding err in ID?

Reply:
@yussufsh 
The id for this function is the cloud instance id appended to every volume id.

The wait function isWaitForIBMPIVolumeBulkDeleted will check that every volume is actually deleted. Every volume that is not deleted will be retried here:


```
                        volumesDelete := models.VolumesDelete{
				VolumeIDs: leftoverVolumeIDs,
			}
			_, err := client.BulkVolumeDelete(&volumesDelete)
			if err != nil {
				var leftoverIDs string
				for _, leftoverVolumeID := range leftoverVolumeIDs {
					leftoverIDs += "/" + leftoverVolumeID
				}
				leftoverError := errors.New(leftoverIDs)
				return leftoverVolumeIDs, State_Error, leftoverError
			}
			return leftoverVolumeIDs, State_Deleting, nil
```

If it times out or errors during this process it returns the leftover undeleted volume ids in the error field appended together, and then sets the id for the resource to the cloud instance id plus this list.

which is this:
```
d.SetId(cloudInstanceID + err.Error())
```

Reply:
The goal was to try to keep track the remaining volumes in case of an error cleaning up though I think this would probably would end up tainting the resource anyway, so I'm not 100% sure how useful this is.